### ISSUE TITLE
Request: Heat map

### ISSUE THREAD

**Issue by andrew-davie** [stella-emu/stella#586]
OK, I am fairly certain this won't meet with approval...

I'd like to have a memory heat-map available.  Every time stella accesses any memory location, it increments a counter for that location. This would be very interesting for a code archaeology viewpoint, but also provide some useful diagnostics.

For example, I could run the program, then dump the heat map and look at the stack area. I could see how much stack was required and if the biggest user was a one-off or happened frequently.  I could watch zero page locations and see those that are never used. 

I could have a heat map of code and particularly subroutines and figure out which need to be optimised first (i.e., those most frequently called). Well, maybe.

Anyway, most of all I thought it would make lovely artwork - each game to have a visual showing the memory usage of the game after it has been played. Have I seen this somewhere before?  I don't know, but it rings a bell now I think about it.

**Comment by thrust26** [comment]
Would that be for RAM, ROM or both? How about hardware addresses?

**Comment by andrew-davie** [comment]
All would be useful. I see some sort of CSV file output, and then any sort of data visualiser could be used to display sections of RAM, or time-baed usage, whatever.

**Comment by thrust26** [comment]
Stella currently tracks address access, but only as a boolean and without considering banks. The latter has to be fixed first (this will require some refactoring), then the former will become relatively easy.

**Comment by thrust26** [comment]
Before we extend the `PageAccess` struct, it needs some more fundamental enhancement first.

Currently it only tracks access within one common 8K structure (2x64 64 bytes pages). The lower 4K are for device access and the upper 4K for code access. So no support for bankswitched ROMs, all code access is tracked within the latter 4K. Finally the address space is limited to 16 bit, which will not be enough for future homebrews (e.g. Penult).

Improving this code will also be a first step for better debugging in bankswitched ROMs (DiStella!). But before we start enhancing it, we should discuss how to proceed here. E.g. how should we refactor the `PageAccess` struct? And should we split device and code access into separate structures?

Target would be post 6.1 of course.

**Comment by thrust26** [comment]
Today I made some good progress with enhancing 'savedis' for multi-bank ROMs. However the PageAccess still needs to be enhanced.

**Comment by sa666666** [comment]
I think this is going to be more involved.  I was thinking that the disassembly logic for each scheme probably needs to be extracted into their respective `CartXXXWidget` classes, since some cart are disassembled differently, some don't have banks in the normal sense, etc.  Certainly some classes of carts are done the same way (ie, `CartF4/F6/F8`, etc), but there are also some others that are quite different.

I think going forward, this needs to be broken out of `CartDebug`, which is getting a little too large and unwieldy.

**Comment by thrust26** [comment]
Agreed. And I already added a note to the code. But IMO we should use this as a base and go for `PageAccess` first.

BTW: I wonder why some Cart classes explicitly add their RAM to `createRomAccessBase`. IMO RAM should be handled differently than ROM in any case.

**Comment by sa666666** [comment]
I think it's because for some schemes, RAM is sometimes active and sometimes in a quiescent state.  So it needs to track the RAM and the address space the RAM would occupy separately.  At least that's all I can remember; it's been a long time.

**Comment by sa666666** [comment]
> Before we extend the `PageAccess` struct, it needs some more fundamental enhancement first.
> 
> Currently it only tracks access within one common 8K structure (2x64 64 bytes pages). The lower 4K are for device access and the upper 4K for code access. So no support for bankswitched ROMs, all code access is tracked within the latter 4K. Finally the address space is limited to 16 bit, which will not be enough for future homebrews (e.g. Penult).
> 
> Improving this code will also be a first step for better debugging in bankswitched ROMs (DiStella!). But before we start enhancing it, we should discuss how to proceed here. E.g. how should we refactor the `PageAccess` struct? And should we split device and code access into separate structures?
> 
> Target would be post 6.1 of course.

We currently create an array the size of the ROM for determining the type of access (`romAccessBase`).  It is mapped to each page, and the array contains 16-bit values).  So for example, if a ROM is 64KB, then a 64KB array of uInt16 (aka, `Device::AccessFlags`) is created.  It is mapped to each page as necessary.

I guess we could add another pointer to `PageAccess`, which again is sized according to the size of the ROM, but this time make it 32-bit.

Remember, the code/data for a ROM, the data for `Device::AccessFlags` for each byte of ROM, and the new array containing the 'heatmap data' exists independently of the `PageAccess` structure.  The `PageAccess` structure simply contains pointers _into those arrays for that particular page_.  Code access is tracked within the 4K window for sure, but in a multi-bank ROM that 4K window points to various pages at different times, and when the page changes, so does the pointer to the specific part of `romAccessBase`.  So the infrastructure is already there; just add another array, and another pointer to `PageAccess`, and the paging will take care of swapping in and out when necessary.

**Comment by thrust26** [comment]
> I guess we could add another pointer in PageAccess, which again is sized according to the size of the ROM, but this time make it 32-bit.
> ...just add another array, and another pointer to PageAccess,

Can you explain that, please? What should that extra pointer (in or to?) PageAccess do?

> _The PageAccess structure simply contains pointers into those arrays for that particular **page**._

I suppose you mean _**bank**_ here?

**Comment by thrust26** [comment]
> I think this is going to be more involved. I was thinking that the disassembly logic for each scheme probably needs to be extracted into their respective `CartXXXWidget` classes, since some cart are disassembled differently, some don't have banks in the normal sense, etc. Certainly some classes of carts are done the same way (ie, `CartF4/F6/F8`, etc), but there are also some others that are quite different.
> 
> I think going forward, this needs to be broken out of `CartDebug`, which is getting a little too large and unwieldy.

Agreed. But currently the `CartXXX` and `CartXXXWidget` classes are mostly copy and paste. IMO we should first refactor them, so that we do not have to create more redundant code.

**Comment by sa666666** [comment]
I think we need to move the discussion of disassembly and related refactoring to another issue.  Perhaps we need to reopen https://github.com/stella-emu/stella/issues/184, and move all multi-bank disassembly discussion there.  That issue was never fully resolved, so it probably shouldn't have been closed.

And this issue is about tracking accesses to RAM and ROM, which is only tangentially related to disassembly.

**Comment by thrust26** [comment]
Agreed for the separate issue. But maybe better a new more general one.

**Comment by sa666666** [comment]
I leave it to you to decide on the issue name.

**Comment by thrust26** [comment]
So for implementing this request, we need a counter in the `PageAccess` struct.

**Comment by sa666666** [comment]
We need a counter _for every possible byte in the address space_.  So we need an array of counters.  That array would be created elsewhere, similar to the one that exists for tracking `Device::AccessFlags` for each byte of ROM.

The `PageAccess` struct would have a _pointer_ into that array, specific for that particular page of ROM/RAM/whatever.

So several things are needed:
- create an array of counters, similar to `Cartridge::myRomAccessBase`
- add a pointer to `PageAccess` that points to this array, similar to `System::PageAccess::romAccessBase`
- initialize the previously created pointer similar to how `Cartridge::myRomAccessBase` is initialized (ie, in each `CartridgeXXX::bank()` method)
- on each System::peek, update that pointer

Really, this is not much different than how we track the type of ROM.  So just follow the same logic/methods/etc, but instead of setting bitflags, increment a counter.  It's exactly the same idea.

**Comment by thrust26** [comment]
Agreed, I wasn't precise here.

**Comment by thrust26** [comment]
@andrew-davie Should we separately count read and write accesses?

**Comment by sa666666** [comment]
And if we do, we need to double the space we allocate.  I assume 16-bits (and counts of 65536) aren't large enough, so we need to use 32 bits?

I suggest only creating one array, so we only have to add one pointer.  The reads can be in the lower half of the array, and the writes in the upper half.

**Comment by thrust26** [comment]
I think 32 bits might be too small too. For now I am aiming at 64 bit (using a dedicated type, so this can be changed easily any time).

**Comment by sa666666** [comment]
Do you really think that we will have over 4 billion accesses to an address?  I suppose using 64-bit is fine, and not any issue (other than a potential waste of memory).

**Comment by andrew-davie** [comment]
Yes. For example, in some of my 3E code, I have a 2K ROM bank defined. The first 1K of that ROM bank is copied to RAM and I both execute code in that RAM, but also store variables in that RAM. Variable writes into that RAM have +$400 added to the address. So, $F000-$F3FF is the read address (Code and RAM), and $F400-$F7FF is the write address. But sometimes I don’t use the RAM, and just copy the whole 2K of ROM to Ithe $F000-$F7FF address. And I execute that code.
So there IS a distinct requirement to differentiate between reading and writing. Writing to $F400… will be a RAM location at $F000…. And Reading $F400... is definitely access to code.
So, I’m thinking either
1) keep track of read and write accesses
Or 2) build in an understanding that writing F400 (for example) is *actually* writing F000
Another way of saying 2) is that F400 isn’t REALLY F400. It is F000 (when writing).
Cheers
A

**Comment by thrust26** [comment]
@sa666666 
Usually all devices handle their page accesses on their own. But for the `Cartridge` device this is split between `System` and `Cartridge` device.

Why is this different? Wouldn't it be better to move all code into the `Cartridge` class?

**Comment by sa666666** [comment]
Not sure I understand the question.  The `System` has the concept of pages, not exclusively a `Cartridge`.  All peeks and pokes go through `System`, which then passes them along to subsystems.

Sometimes a cart has to access pages outside it's own address space.  With mirroring, sometimes the cart needs to handle writes to TIA/RIOT/etc.  So it takes responsibility for them.

Also, I assume you know there is more work to be done.  The line you have in `Cart4K` (`access.romAccessCounter = &myRomAccessCounter[addr & 0x0FFF];`) needs something similar in many other classes, otherwise the following code in `System::peek` will always call a method instead of accessing the array directly, leading to potential slowdown:
```
  // Increase access counter
  if(access.romAccessCounter)
    *(access.romAccessCounter + (addr & PAGE_MASK)) += 1;
  else
    access.device->increaseAccessCounter(addr);
```

**Comment by sa666666** [comment]
Also, I believe the extension for the generated file should be 'csv', not 'cvs' (**c**omma **s**eparated **v**alues). :smiling_imp:

**Comment by thrust26** [comment]
> Not sure I understand the question. The System has the concept of pages, not exclusively a Cartridge. All peeks and pokes go through System, which then passes them along to subsystems.

Right. But for the cart (unlike `TIA` and `M6532`) the access updates are done in `System` instead of `Cartridge`. This seems non consistent.

> otherwise the following code in System::peek will always call a method instead of accessing the array directly, leading to potential slowdown:

That answers another question regarding the special handling for cart access. But is that really necessary? And if there is noticeable slowdown, wouldn't it be better to let the compiler make the optimizations?

> Also, I believe the extension for the generated file should be 'csv', not 'cvs' (comma separated values).

Good catch! :smile:

**Comment by sa666666** [comment]
For the first point, sorry, still not seeing what you mean.  Perhaps you could show a code snippet or two illustrating this, and how you think it should work.

This design pattern (first access by pointer, and if not possible use a method call) is pretty fundamental in how Stella works.  It's been there since the beginning, and I assume it makes (or made) a difference.  Stella started when 486 machines were common, so every bit of speed was necessary.

As for the compiler, if we remove this infrastructure, there is no _optimization_ for the compiler to make.  Right now, it's either (a) access a pointer directly, or (b) call a virtual function.  If we remove (a), then we _have_ to do (b).  And since the method is virtual, it can't be inlined.  So it still ends up being a pointer dereference (for the method name), and then a jump/return (for the method call).  It's hard to see how that can ever be faster than a simple pointer dereference.

In any event, I'm not opposed to doing research on this, and reworking Stella to do things differently.  But I don't want to do it as part of this issue, and definitely not in master.  So if we want to pursue that, perhaps make it another issue and work on it in another branch.  Modifying this is making a fundamental design change, and not something I want in master until it would receive extensive testing.

**Comment by thrust26** [comment]
> For the first point, sorry, still not seeing what you mean. Perhaps you could show a code snippet or two illustrating this, and how you think it should work.

If you look at e.g. `TIA`, then you see that the e.g. the `getAccessFlags` method is implemented there. But for `Cartridge` it is implemented in `System`.

> Right now, it's either (a) access a pointer directly, or (b) call a virtual function. If we remove (a), then we have to do (b). And since the method is virtual, it can't be inlined. 

I understand. Not sure if that's really relevant anymore (especially because it is only for platforms with debugger support), but I understand that this requires testing first.

However, the resulting code for the access counter in `System` will either need 
- access to `Cartridge` to get the image size (offset for writes into the array), or
- require separate arrays for read and write access, or 
- interleaved storing of read and write accesses 

I suppose separate arrays is the best option here. Do you agree?

**Comment by sa666666** [comment]
> If you look at e.g. `TIA`, then you see that the e.g. the `getAccessFlags` method is implemented there. But for `Cartridge` it is implemented in `System`.

The one in `TIA` is specific for the TIA.  The one in `System` is a 'gatekeeper' that arbitrates calls to all the other devices.  That way, other parts of the code can call `System::getAccessFlags`, and it will route the call to the correct device, **including** making a call to `TIA::getAccessFlags`, if necessary.

It doesn't make sense to move the code to `Cartridge` directly, as `System` is the main control class that **owns** devices (TIA and Cart included).  You'll also note that a `Cartridge` can have its own `getAccessFlags` too, if necessary.  `Cart4A50` is one that does.  So perhaps you're confused with the naming.  The method `System::getAccessFlags` only has the same name as the others, but not because of inheritance or anything like that.  You could change that name, and _still_ need `getAccessFlags` in other `Device` classes.

> I understand. Not sure if that's really relevant anymore (especially because it is only for platforms with debugger support), but I understand that this requires testing first.

If you get rid of the 'pointer vs. virtual call' design, you need to do it for the peek and poke bases as well.  And that definitely affects all classes, not just when debugging support is included.  And if you get rid of that, you also don't need paging anymore.  So that's another upheaval.  As I said before, I think the paging/pointer to arrays design was to make things as fast as possible.  Do we still need it?  I would argue we might, for R77 and older systems in particular.  And if we change it, it requires deep surgery to most of the emulation core classes in Stella.  All `Device` classes, `System`, etc.  I wonder if it's worth the effort to change all this, and still potentially end up with slower emulation.

> However, the resulting code for the access counter in `System` will either need
> 
> * access to `Cartridge` to get the image size (offset for writes into the array), or

No, you just set up the paging correctly in each `Cart` class, and that takes care of it.  So take `Cart4KSC`, which is probably the simplest class that has RAM:
```
  access.type = System::PageAccessType::WRITE;
  for(uInt16 addr = 0x1000; addr < 0x1080; addr += System::PAGE_SIZE)
  {
    access.romAccessBase = &myRomAccessBase[addr & 0x007F];
    mySystem->setPageAccess(addr, access);
  }

  // Set the page accessing method for the RAM reading pages
  access.type = System::PageAccessType::READ;
  for(uInt16 addr = 0x1080; addr < 0x1100; addr += System::PAGE_SIZE)
  {
    access.directPeekBase = &myRAM[addr & 0x007F];
    access.romAccessBase = &myRomAccessBase[0x80 + (addr & 0x007F)];
    mySystem->setPageAccess(addr, access);
  }

  // Map ROM image into the system
  for(uInt16 addr = 0x1100; addr < 0x2000; addr += System::PAGE_SIZE)
  {
    access.directPeekBase = &myImage[addr & 0x0FFF];
    access.romAccessBase = &myRomAccessBase[addr & 0x0FFF];
    mySystem->setPageAccess(addr, access);
  }
```
Note how in the read and write areas, `access.romAccessBase` is mapped to different parts of `myRomAccessBase` (in particular, reads are offset by $80)?  You just need to do a very similar thing, except use the variables you created for the counters.  And since you're doing it in the `Cart` class, it already knows its own size, and doesn't need to expose that to any other class.

**Comment by sa666666** [comment]
Wait, I also thought of something else.  You're tracking read and write counts, so you need to access the single pointer in two different ways.  I didn't realize that.  So perhaps you could add two pointers to `PageAccess` instead, and model how `directPeekBase` and `directPokeBase` works.  Still one array behind the scenes, but two pointers that point to different parts of it.

**Comment by thrust26** [comment]
> The one in TIA is specific for the TIA. 

Correct. And the one in `M6532` is specific for the RIOT. Why not following the same pattern for `Cartridge`? ROM and extended RAM accesses are specific for it.

> The one in System is a 'gatekeeper' that arbitrates calls to all the other devices. That way, other parts of the code can call System::getAccessFlags, and it will route the call to the correct device, including making a call to TIA::getAccessFlags, if necessary.

I wouldn't want to remove that 'gatekeeper' functionality because it is useful. But there are many situations where we can use the virtual function call or directly call the device. The latter is currently not possible for `Cartridge`, we have to use `System` instead.

> If you get rid of the 'pointer vs. virtual call' design, you need to do it for the peek and poke bases as well. 

Here I am most likely missing something. I suppose you are assuming the `directpeek/pokebase` would be affected too. This was never my intention. 

Probably you just want to keep things consistent. Basically the same as I want, just from a different perspective?

> So it still ends up being a pointer dereference (for the method name), and then a jump/return (for the method call). It's hard to see how that can ever be faster than a simple pointer dereference.

Plus an `if`, but that's of course much cheaper than a method call. :smile:

**Comment by sa666666** [comment]
OK, some history.

In the beginning, all debugger access was through `romAccessBase` pointer.  However, certain classes needed more fine-grained access, since pages are 64-bytes and sometimes we need byte access.  So the `getAccessFlags` method was added as `virtual` to `Device`.  And then only `Cart4A50` and `CartAR` need it.

So, you can still have that method for every `Cart` class if you like, since it's a `virtual` method that can be overrided.  But the point was to use the paging system instead, for reasons mentioned above.  So if we fix the mapping inside each cart class, we **don't need** to use `getAccessFlags` at all.  Either way, we're going to have to modify every cart class.  Either we modify those classes to set up the paging correctly, or we modify them to include all those extra methods.  I prefer the former, since it's probably faster, and definitely less code.

Also, you can call `getAccessFlags` on the cart directly.  It's just that for most carts, it will call a default `virtual` method that doesn't do anything.  So there's no need to go through `System` to call that method.  What I'm saying is that if we set up the mapping/paging correctly, you almost never need to call that method.

**Comment by sa666666** [comment]
The current design is basically use one or the other of:
- paging, and use the pointers directly to the arrays
- `getAccessFlags` and `setAccessFlags`

These are not meant to be used at the same time.  So most of the time we use the former, and only in cases where required do we use the latter.

**Comment by sa666666** [comment]
Furthermore, the way paging is set up is:
- if `directPeekBase`, `directPokeBase` and friends are non-null, use the pointer directly
- if they are null, then call `peek` and `poke` directly

Some classes are very straightforward, and can do everything through pointers.  So their `directPeekBase` and `directPokeBase` are set directly, and their `peek` and `poke` are either empty, or handle exceptional cases only (like read from write port, etc).

Other classes are much more complicated (like `Cart4A50`), so the paging is set such that **all** accesses are mapped directly to their `peek` and `poke` (and then their `directPeekBase` and `directPokeBase` are set to null).

The debugger stuff is modelled in the same way.  Set the pointers for direct access, or set them to null and use `getAccessFlags` and `setAccessFlags` instead.  See `Cart4A50` for more explanation.

**Comment by thrust26** [comment]
Thanks for the explanation.

>  So if we fix the mapping inside each cart class, we don't need to use getAccessFlags at all. 

Not sure what you want to 'fix' here. You mean `setup' (like we do it now), right?

> Either way, we're going to have to modify every cart class. Either we modify those classes to set up the paging correctly, or we modify them to include all those extra methods. 

That wouldn't be necessary. They are all derived from the `Cartridge` class, so except for the few existing exceptions, they would use their parent's class code. But I suppose I will soon find out why this wouldn't work. :smile:

BTW: PageAccessType seems to be set but (except for MNetwork) never read or checked. What am I missing here?

**Comment by sa666666** [comment]
> BTW: `PageAccessType` seems to be set but (except for MNetwork) never read or checked. What am I missing here?

It's possible that I intended to use it for the debugger, but never did get around to it.  Or maybe it was added in anticipation of refactoring the cart classes, but again never finished it.  After the rework, if we find it isn't used, I guess it can be removed.  But we should probably leave it alone for now.

**Comment by sa666666** [comment]
Regarding `PageAccessType`, you've definitely found a use for it when you refactored the `MNetwork` classes.  So perhaps it will be needed in a similar way when refactoring other cart classes.

**Comment by thrust26** [comment]
@andrew-davie I think with 8b550a3 I have implemented enough functionality for you request. Can you test the current version, please?

**Comment by thrust26** [comment]
@sa666666 
BTW: `CartFE` misses the access code, thus all peeks (and pokes) result into virtual method calls. 

For the same reason the disassembly is very limited.

**Comment by sa666666** [comment]
> @sa666666
> BTW: `CartFE` misses the access code, thus all peeks (and pokes) result into virtual method calls.
> 
> For the same reason the disassembly is very limited.

Copy/paste error I guess.  So it's working fine now?

**Comment by andrew-davie** [comment]
> @andrew-davie I think with [8b550a3](https://github.com/stella-emu/stella/commit/8b550a3a11beafcde829084c14d915307e4db05b) I have implemented enough functionality for you request. Can you test the current version, please?

Thanks Thomas. I've tried to give it a go on a version of Qb...

[Qb (2.15) (Retroactive) (NTSC).csv.zip](https://github.com/stella-emu/stella/files/4421612/Qb.2.15.Retroactive.NTSC.csv.zip)

I haven't had much more than a cursory look at this stage, but it seems to me that there is an error in the RAM reads to start with. If I'm understanding the format correctly...

RAM reads:
0080,  385869, 0081,       0, 0082, ...

3.8 million reads to locatino 0x80 and 0 to all of the other RAM.
In fact, looks like some of the "read" counters are borked - giving the first location with total reads, but all the other locations zero.  Stack, for example...

Stack reads:
0180,  236615, 0181,       0, 0182,       0, 0183,       0, 0184,       0, 0185,       0, 0186,   ....


But this is a great start - and thank you.
I'm tempted to knock up a little python program to display the heat map!

**Comment by andrew-davie** [comment]
This may end up being VERY useful for debugging.
I know it's preliminary and may be buggy, but consider the following...

```
RAM writes:
0080,    3724, 0081,    3064, 0082,    2368, 0083,    2477, 0084,    2636, 0085,    2146, 0086,    1842, 0087,    1842, 0088,    2432, 0089,    3932, 008a,    3782, 008b,    1476, 008c,       0, 008d,       0, 008e,    1840, 008f,     319, 0090,   14494, 0091,    2006, 0092,    7024, 0093,    2879, 0094,    1654, 0095,    1404, 0096,    1482, 0097,    1380, 0098,    6791, 0099,   13814, 009a,    2852, 009b,    2852, 009c,    2945, 009d,    2945, 009e,    3470, 009f,    3470, 00a0,    3190, 00a1,    3190, 00a2,    5522, 00a3,    5520, 00a4,    3047, 00a5,    3147, 00a6,    3360, 00a7,    3889, 00a8,    8937, 00a9,    8946, 00aa,    8955, 00ab,    8964, 00ac,    8315, 00ad,    8787, 00ae,    8782, 00af,    8777, 00b0,    8772, 00b1,    8295, 00b2,    8649, 00b3,    8646, 00b4,    8643, 00b5,    8640, 00b6,    8283, 00b7,    8558, 00b8,    8557, 00b9,    8556, 00ba,    8555, 00bb,    8937, 00bc,    8946, 00bd,    8955, 00be,    8964, 00bf,    8315, 00c0,    8787, 00c1,    8782, 00c2,    8777, 00c3,    8772, 00c4,    8295, 00c5,    8649, 00c6,    8646, 00c7,    8643, 00c8,    8640, 00c9,    8283, 00ca,    8558, 00cb,    8557, 00cc,    8556, 00cd,    8555, 00ce,    8937, 00cf,    8946, 00d0,    8955, 00d1,    8964, 00d2,    8315, 00d3,    8787, 00d4,    8782, 00d5,    8777, 00d6,    8772, 00d7,    8295, 00d8,    8649, 00d9,    8646, 00da,    8643, 00db,    8640, 00dc,    8283, 00dd,    8558, 00de,    8557, 00df,    8556, 00e0,    8555, 00e1,    3210, 00e2,    9675, 00e3,    3079, 00e4,    2005, 00e5,    2004, 00e6,    2004, 00e7,  132358, 00e8,  131048, 00e9,   41662, 00ea,   40757, 00eb,   45502, 00ec,   51345, 00ed,  135501, 00ee,  128360, 00ef,   37751, 00f0,   35940, 00f1,   57569, 00f2,   28980, 00f3,    1380, 00f4,   20700, 00f5,   44160, 00f6,    2760, 00f7,  164220, 00f8,  138000, 00f9,  131100, 00fa,       0, 00fb,       0, 00fc,       0, 00fd,       0, 00fe,       0, 00ff,       0, 
```

That's the stack up there at the "top" I would have thought, and I most certainly would have expected to see things in use from 0xFA-0xFF, so if that's still showing like that when this mod is "in production", I'll have some interesting review/hunting to do!

**Comment by thrust26** [comment]
The stack accesses are listed separately, directly below the RAM accesses.

I will look into the RAM reads now. Thanks for testing!

**Comment by andrew-davie** [comment]
Thank you. The stack reads are broken too!

**Comment by andrew-davie** [comment]
This is great. At one glance I can see that the stack writes never go below 0xFB and the RAM writes never go above 0xFA.  Gives me some confidence at a glance that I got my stack/RAM cooperation correct. Well, I  *know* i did, but it's nice to look at actual usage and see that it's working as expected :)

**Comment by thrust26** [comment]
@andrew-davie I fixed the RAM/Stack access with 17835be. 

Note: stack pops were not traced before.

**Comment by thrust26** [comment]
I suppose we can close this now.
